### ISSUE TITLE
 Pattern Matching for Switch Fails in Java 21 

### ISSUE THREAD

**Issue by volosied** [eclipse-jdt/eclipse.jdt.core#1574]
For the code (used within the context of a JSP) below: 

```
       Object test = "hello";
       switch(test){
         case Integer i -> out.println("switch fail!");
         case String s ->  out.println("switch success!");
         default  -> out.println("switch fail!");
       }
```
     
JDT ( `org.eclipse.jdt:org.eclipse.jdt.core:3.35.0` )  encounters errors such as: 
-  Cannot switch on a value of type Object. Only convertible int values, strings or enum variables are permitted
-  Syntax error on token "i", delete this token
-  String cannot be resolved to a variable

CompilerOptions.OPTION_Source, CompilerOptions.OPTION_Compliance, CompilerOptions.OPTION_TargetPlatform are all set to `21`. 

I don't understand why.  If test the code above using Jshell (and modify it to use System.out) then it works fine.  It seems like a bug, in my opinion.

**Comment by volosied** [comment]
Looks like a few defects in this area have already been opened: 

https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1466
https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1558

**Comment by volosied** [comment]
Looks to be fixed in the snapshot: `ecj-3.36.0-20231103.101317-173.jar` 

Closing!
